[7.0.x] service state

a-palchikov opened the pull request:
## Description
<!--Required. Provide high-level overview of what the change is for.-->
This PR fixes the installer/expand/upgrade operations on CentOS 8/RHEL by moving the gravity binary to the system location intended for binary files. This is important on SELinux-enabled systems where `init_t` (systemd) domain does not have permissions to work with executables in arbitrary locations (e.g. user home directories or `/var/lib/gravity`).
This prompted a series of changes that mainly concern the agent service's operational state to maintain it the same way it was maintained previously. The main problem is supporting both installer/agent services on a single box (when the installer is running in `remote` mode - i.e. does not itself represent a joining agent) which requires that the state is separate for each.

## Type of change
<!--Required. Keep only those that apply.-->

* Bug fix (non-breaking change which fixes an issue)
* Internal change (not necessarily a bug fix or a new feature)
* This change has a user-facing impact

## Linked tickets and other PRs
<!--Required. Keep only those that apply.-->

<!--This PR addresses the following issues.-->
* Refs https://github.com/gravitational/gravity/issues/2009
* Requires https://github.com/gravitational/gravity.e/pull/4345

## TODOs
<!--Required. Keep only those that apply and check them off as they get completed.-->

- [x] Self-review the change
- [x] Perform manual testing (see below)
- [ ] Write documentation
- [ ] Address review feedback

## Implementation
<!--Optional. Add any relevant implementation details that might help the reviewers.-->

## Performance/Scaling
<!--Optional. Add any relevant details on how this PR reacts when scaled to 1k nodes, and any additional scaling considerations for the reviewers.-->

## Testing done
<!--Required. Explain what kind of testing these changes underwent.-->
 * [x] Installs on a single node
 * [x] Expands
 * [ ] Node reconfiguration
 * [ ] Upgrades

## Additional information
<!--Optional. Anything else that may be relevant.-->

a-palchikov committed:
Now that the install/join agents are running from a binary that resides
in a known binary location (/usr/bin/ or /usr/local/bin) and local operation state
is still kept in a working directory of tthe client, we need a way to
let the service process know about the state directory.

Introduce a new CLI argument where necessary to convey this
configuration and refactor handling of state for services.

a-palchikov committed:
Explicit socket/service path configuration for the resume strategy.

a-palchikov committed:
Remove excessive dependency on the state directory - instead provide
functions with the information they would otherwise search in state
directory directly.

r0mant commented on tool/gravity/cli/operation.go:
```diff
@@ -508,7 +510,8 @@ func (r *backendOperations) listUpdateOperation(environ LocalEnvironmentFactory)
 
 func (r *backendOperations) listJoinOperation(environ LocalEnvironmentFactory) error {
-	env, err := environ.NewJoinEnv()
+	// TODO(dmitri): assumes local state inside process's working directory
```
What does this mean? Join-related plan commands have to be run from a specific directory? Is that a change in behavior?

a-palchikov replied:
No, this is not a change in behavior - it is note to self when we need to change this behavior.

r0mant commented on lib/state/state.go:
```diff
@@ -144,12 +144,40 @@ func LogDir(baseDir string, suffixes ...string) string {
 
+// GravityInstallerServicePath returns the path to the installer service
+// as configured in the state directory rooted at baseDir.
+func GravityInstallerServicePath(baseDir string) (path string, err error) {
```
Is this function actually used anywhere? Couldn't find it.

a-palchikov replied:
Yes, it is used [here](https://github.com/gravitational/gravity/blob/dmitri/7.0.x/2009-service-state/tool/gravity/cli/install.go#L954)

r0mant commented on tool/gravity/cli/install.go:
```diff
@@ -683,6 +704,10 @@ func completePlanFromService(
 	operation ops.SiteOperation,
 	connecting, connected string,
 ) error {
+	strategy, err := newResumeStrategy(utils.Exe.WorkingDir)
```
Could you please explain a bit what kind of impact this change has on UX? Are all of these commands now depend on the working dir of the binary somehow?

I.e. will users still be able to just execute `gravity join`, `gravity plan`, etc. commands from wherever on cluster nodes?

a-palchikov replied:
No, I'm simply pulling the previously implicit state configuration to the top so it's easier to change if necessary - there is no change in behavior.

r0mant commented on tool/gravity/cli/register.go:
```diff
@@ -129,6 +130,7 @@ func RegisterCommands(app *kingpin.Application) *Application {
 	g.AutoJoinCmd.Token = g.AutoJoinCmd.Flag("token", "Unique token to authorize this node to join the cluster.").Hidden().String()
 	g.AutoJoinCmd.SELinux = g.AutoJoinCmd.Flag("selinux", "Run with SELinux support. Default 'false'.").Default("false").Envar(defaults.GravitySELinuxEnv).Bool()
 	g.AutoJoinCmd.FromService = g.AutoJoinCmd.Flag("from-service", "Run in service mode.").Hidden().Bool()
+	g.AutoJoinCmd.StateDir = g.AutoJoinCmd.Arg("path", "Path to the directory with local state. Defaults to working directory.").String()
```
What are these flags for? How does this change behavior for existing users (if it does)?

a-palchikov replied:
These flags specify to the service where the operation-specific state is.
Now that I'm thinking about it, we can continue using the process's working directory for the service since this is how we set up installer/join agents in the first place.

a-palchikov commented on lib/defaults/defaults.go:
```diff
@@ -371,6 +374,12 @@ const (
 	// StatBin is stat executable path inside planet
 	StatBin = "/usr/bin/stat"
 
+	// AlternativeBinDir defines the default location for binaries on Ubuntu Core
```
This will simplify things but not much. On the good side, we'll be throwing away code - not changing it.

a-palchikov committed:
Address review comments

a-palchikov commented:
It looks like using process's working directory does not work in all cases  specifically when the operation-specific phases are remotely executed by individual calls to `gravity plan execute`. The plan now is to attempt and move operation-specific state to a fixed location (probably under state directory) so this needs a bit more work.

a-palchikov committed:
Fix reconfiguration workflow (it is still missing service/pod subnet config though). Start self RPC commands using the current process's working directory

a-palchikov committed:
lib/update/cluster: remove explicit package unpacking during bootstrap
step.

Bootstrap step will explicitly unpack all update packages into their
default location incl. the planet package. If the planet package did not
change, the installed and updated packages will be the same and the
unpacking will overwrite volatile container's configuration like the
resolv.conf.
Do not unpack the package to avoid overwriting. Figure out another way
to deal with https://github.com/gravitational/gravity/issues/1279 which
this change was introduced for.

a-palchikov committed:
Download agent binary to /usr/local/bin by default in interactive installation to avoid collision with the installer on single node

r0mant commented on lib/ops/opsservice/instructions.go:
```diff
@@ -36,32 +36,33 @@ var (
 	// gravityTemplateSource is a bash script that downloads gravity binary
 	// from an Ops Center and installs it into /usr/bin
 	gravityTemplateSource = `
+
```
Is this an empty template now?

a-palchikov replied:
Forgot to remove it.

r0mant reviewed (approved):
lgtm if operations are working as expected

a-palchikov committed:
lib/system/service: instead of stopping/disabling previous service,
uninstall it when reinstalling a service.

This will make sure that the unit file is also removed when possible and
at the correct moment in the whole flow.

a-palchikov committed:
Determine the correct binary to use for 'plan resume' depending on the
operation type.

a-palchikov commented:
Going to merge as the failures seem to be unrelated to robotest (which ran fine).
